r2-shared/src/main/java/org/readium/r2/shared/format/MediaType.kt

```diff
@@ -0,0 +1,307 @@
+ * Use of this source code is governed by a BSD-style license which is detailed in the
+ * LICENSE file present in the project repository where this source code is maintained.
+ */
+
+package org.readium.r2.shared.format
+
+import org.readium.r2.shared.publication.Link
+import java.lang.IllegalArgumentException
+import java.nio.charset.Charset
+import java.util.*
+
+/**
+ * Represents a string media type.
+ *
+ * [MediaType] handles:
+ *  - components parsing – eg. type, subtype and parameters,
+ *  - media types comparison.
+ *
+ * Comparing media types is more complicated than it looks, since they can contain parameters,
+ * such as `charset=utf-8`. We can't ignore them because some formats use parameters in their
+ * media type, for example `application/atom+xml;profile=opds-catalog` for an OPDS 1 catalog.
+ *
+ * Specification: https://tools.ietf.org/html/rfc6838
+ *
+ * @param string String representation for this media type.
+ */
+class MediaType private constructor(string: String) {
+
+    /** The type component, e.g. `application` in `application/epub+zip`. */
+    val type: String
+
+    /** The subtype component, e.g. `epub+zip` in `application/epub+zip`. */
+    val subtype: String
+
+    /** The parameters in the media type, such as `charset=utf-8`. */
+    val parameters: Map<String, String>
+
+    init {
+        if (string.isEmpty()) {
+            throw IllegalArgumentException("Invalid media type: $string")
+        }
+
+        // Grammar: https://tools.ietf.org/html/rfc2045#section-5.1
+        val components = string.split(";")
+            .map { it.trim() }
+        val types = components[0].split("/")
+        if (types.size != 2) {
+            throw IllegalArgumentException("Invalid media type: $string")
+        }
+
+        // > Both top-level type and subtype names are case-insensitive.
+        this.type = types[0].toLowerCase(Locale.ROOT)
+        this.subtype = types[1].toLowerCase(Locale.ROOT)
+
+        // > Parameter names are case-insensitive and no meaning is attached to the order in which
+        // > they appear.
+        val parameters = components.drop(1)
+            .map { it.split("=") }
+            .filter { it.size == 2 }
+            .associate { Pair(it[0].toLowerCase(Locale.ROOT), it[1]) }
+            .toMutableMap()
+
+        // For now, we only support case-insensitive `charset`.
+        //
+        // > Parameter values might or might not be case-sensitive, depending on the semantics of
+        // > the parameter name.
+        // > https://tools.ietf.org/html/rfc2616#section-3.7
+        //
+        // > The character set names may be up to 40 characters taken from the printable characters
+        // > of US-ASCII.  However, no distinction is made between use of upper and lower case
+        // > letters.
+        // > https://www.iana.org/assignments/character-sets/character-sets.xhtml
+        parameters["charset"]?.let {
+            parameters["charset"] = it.toUpperCase(Locale.ROOT)
```

Comment:
```suggestion
            parameters["charset"] = Charset.forName(it).name()
```
Since charsets may have several aliases, this change should make comparisons more robust. Charset.name is supposed to provide the canonical name.

Reply:
@qnga Thanks, that's a good idea.

Can you maybe just wrap it in a `try` which falls back on the original value? Because it throws an exception if the charset is unknown.

Reply:
Actually don't worry about it, I'll make the change and add a unit test for that.

Reply:
I changed this in the other PR: https://github.com/readium/r2-shared-kotlin/pull/100/commits/08ccb6f60697b3c84487914b5e98245849cbd9d2